File: receptor/hamming.cpp
#include "hamming.h"

// Calcula la cantidad de bits de paridad necesarios para un bloque de Hamming
int calcularR(int m) {
    int r = 1;
    while ((m + r + 1) > (1 << r)) {
        r++;
    }
    return r;
}

// Determina si un número es una potencia de dos
bool esPotenciaDeDos(int x) {
    return x != 0 && (x & (x - 1)) == 0;
}

// Decodifica un bloque recibido, detecta y corrige un posible error de un bit
ResultadoBloque decodificarBloque(const std::vector<int>& recibido, int m, int r) {
    int n = m + r;
    std::vector<int> bloque(n + 1, 0);
    for (int pos = 1; pos <= n; pos++) {
        bloque[pos] = recibido[pos - 1];
    }

    // Calcula el síndrome para identificar la posición del bit erróneo
    int sindrome = 0;
    for (int k = 0; k < r; k++) {
        int p = 1 << k;
        int paridad = 0;
        for (int pos = 1; pos <= n; pos++) {
            if (pos & p) {
                paridad ^= bloque[pos];
            }
        }
        if (paridad != 0) {
            sindrome += p;
        }
    }

    bool corregido = false;
    if (sindrome != 0 && sindrome <= n) {
        bloque[sindrome] ^= 1;
        corregido = true;
    }

    std::vector<int> datos;
    for (int pos = 1; pos <= n; pos++) {
        if (!esPotenciaDeDos(pos)) {
            datos.push_back(bloque[pos]);
        }
    }

    return {datos, corregido, sindrome};
}

// Recorre toda la trama, corrige bloques individuales y reconstruye los datos
std::vector<int> verificarYCorregir(const std::vector<int>& trama, int m, int r,
                                    int longitudOriginal,
                                    int& bloquesCorregidos,
                                    int& bloquesIrrecuperables,
                                    bool& tramaValida) {
    int n = m + r;
    bloquesCorregidos = 0;
    bloquesIrrecuperables = 0;
    tramaValida = (n > 0) && !trama.empty() && (trama.size() % n == 0);

    std::vector<int> datos;
    if (!tramaValida) {
        return datos;
    }

    for (size_t i = 0; i + n <= trama.size(); i += n) {
        std::vector<int> bloque(trama.begin() + i, trama.begin() + i + n);
        ResultadoBloque resultado = decodificarBloque(bloque, m, r);

        if (resultado.sindrome != 0) {
            if (resultado.corregido) {
                bloquesCorregidos++;
            } else {
                bloquesIrrecuperables++;
            }
        }

        datos.insert(datos.end(), resultado.datos.begin(), resultado.datos.end());
    }

    if ((int)datos.size() < longitudOriginal) {
        tramaValida = false;
        return datos;
    }

    datos.resize(longitudOriginal);
    return datos;
}

File: emisor/hamming.py
# Calcula la cantidad de bits de paridad necesarios para un bloque de Hamming
def calcular_r(m):
    r = 1
    while (m + r + 1) > (1 << r):
        r += 1
    return r


# Verifica si un número es una potencia de dos
def es_potencia_de_dos(x):
    return x != 0 and (x & (x - 1)) == 0


# Codifica un bloque de datos usando Hamming con bits de paridad
def codificar_bloque(datos, m, r):
    n = m + r
    bloque = [0] * (n + 1)

    idx = 0
    # Coloca los bits de datos en las posiciones que no son potencias de dos
    for pos in range(1, n + 1):
        if not es_potencia_de_dos(pos):
            bloque[pos] = datos[idx]
            idx += 1

    # Calcula cada bit de paridad según las posiciones que cubre
    for k in range(r):
        p = 1 << k
        paridad = 0
        for pos in range(1, n + 1):
            if pos != p and (pos & p):
                paridad ^= bloque[pos]
        bloque[p] = paridad

    return bloque[1:]


# Codifica todo el mensaje en bloques de Hamming y devuelve la trama completa
def calcular_integridad(bits_mensaje, m):
    """Codifica el mensaje en bloques de Hamming(n, m).

    Retorna (trama, r, longitud_original).
    """
    r = calcular_r(m)
    longitud_original = len(bits_mensaje)

    datos = list(bits_mensaje)
    resto = len(datos) % m
    if resto != 0:
        # Completa el último bloque con ceros para mantener bloques de tamaño m
        datos += [0] * (m - resto)

    trama = []
    for i in range(0, len(datos), m):
        trama.extend(codificar_bloque(datos[i:i + m], m, r))

    return trama, r, longitud_original
